apollo-ios-codegen/Sources/GraphQLCompiler/JavaScript/src/utilities/fieldPolicyDirective.ts

```diff
@@ -0,0 +1,272 @@
+      result.push(directive)
+    }
+  }
+
+  for (const directive of type.astNode?.directives ?? []) {
+    if (directive.name.value === directiveName) {
+      result.push(directive);
+    }
+  }
+
+  if("getInterfaces" in type) {
+    for (const interfaceType of type.getInterfaces()) {
+      const found = fieldPolicyDirectivesFor(interfaceType);
+      if (!found) {
+        continue;
+      }
+
+      for (const foundDirective of found) {
+        var duplicate = false;
+        for (const directive of result) {
+          if (matchDirectiveArguments(directive, foundDirective)) {
+            duplicate = true;
+            break;
+          }
+        }
+
+        if (!duplicate) {
+          result.push(foundDirective);
+        }
+      }
+    }
+  }
+
+  return result
+}
+
+function matchDirectiveArguments(
+  first: DirectiveNode,
+  second: DirectiveNode
+): boolean {
+  return (
+    (first.arguments ?? [])
+      .map((node) =>
+        JSON.stringify([node.name.value, valueFromASTUntyped(node.value)])
+      )
+      .sort()
+      .toString() ===
+    (second.arguments ?? [])
+      .map((node) =>
+        JSON.stringify([node.name.value, valueFromASTUntyped(node.value)])
+      )
+      .sort()
+      .toString()
+  );
+}
+
+function validateKeyArgs(
+  actualField: GraphQLField<any, any>,
+  keyArgs: string[]
+): void {
+  for (const keyArg of keyArgs) {
+    validateKeyArgPath(actualField, keyArg);
+  }
+}
+
+function validateKeyArgPath(
+  field: GraphQLField<any, any>,
+  keyArg: string
+): void {
+  const parts = keyArg.split(".").filter(Boolean);
+  const [argName, ...path] = parts;
+
+  const arg = field.args.find(a => a.name === argName);
+  if (!arg) {
+    throw new GraphQLError(
+      `@fieldPolicy key argument "${keyArg}" does not exist as an input argument of field "${field.name}".`,
+      { nodes: field.astNode }
+    );
+  }
+
+  let curType: GraphQLInputType = base(arg.type);
+
+  for (let i = 0; i < path.length; i++) {
+    const segment = path[i];
+    if (!isInputObjectType(curType)) {
+      throw new GraphQLError(
+        `@fieldPolicy key "${keyArg}" traverses "${segment}" on non-object input type "${String(curType)}".`,
+        { nodes: field.astNode }
+      );
+    }
+    const fieldMap = curType.getFields();
+    const nextField = fieldMap[segment];
+    if (!nextField) {
+      const suggestions = Object.keys(fieldMap).join(", ");
+      throw new GraphQLError(
+        `@fieldPolicy key "${keyArg}" refers to unknown input field "${segment}" on "${curType.name}". Known fields: ${suggestions}`,
+        { nodes: field.astNode }
+      );
+    }
+    curType = base(nextField.type);
+  }
+
+  if (!parts.length || isInputObjectType(curType)) {
+    throw new GraphQLError(
+      `@fieldPolicy key "${keyArg}" must resolve to a leaf input type (scalar/enum), got "${String(curType)}".`,
+      { nodes: field.astNode }
+    );
+  }
+}
+
+function base(type: GraphQLInputType): GraphQLInputType {
```

Comment:
This function is confusing. The name `base` tells me nearly nothing about what this does, and even though it's one line, I'm still not clear what the intention is here. Is this just a forced downcast?